### PR TITLE
Add a timeout mechanism for the `SocketChannel.open` in the `SslHelper.validateEndpoint` helper

### PR DESCRIPTION
**Pull request by guizmaii**

### CODE REVIEW

File: zio-kafka-test/src/test/scala/zio/kafka/utils/SslHelperSpec.scala [hunk 1]
```diff
@@ -221,7 +221,15 @@ object SslHelperSpec extends ZIOSpecDefaultSlf4j with KafkaRandom {
 
           // We simulate that the Node takes to much time to answer
           val result =
-            SslHelper.doValidateEndpoint { _ => Thread.sleep(2.second.toMillis); ??? }(
+            SslHelper.doValidateEndpoint { _ =>
+              var interrupted = false
+              while (!interrupted) {
+                Thread.sleep(1.second.toMillis)
+                interrupted = Thread.currentThread.isInterrupted
+                println(s"interrupted: $interrupted")
```
Comment (guizmaii) [inline zio-kafka-test/src/test/scala/zio/kafka/utils/SslHelperSpec.scala:229]: TO Clean

File: zio-kafka-test/src/test/scala/zio/kafka/utils/SslHelperSpec.scala [hunk 2]
```diff
@@ -236,6 +244,28 @@ object SslHelperSpec extends ZIOSpecDefaultSlf4j with KafkaRandom {
                 )
             )
           )
+        },
+        test("TODO JULES") {
```
Comment (guizmaii) [inline zio-kafka-test/src/test/scala/zio/kafka/utils/SslHelperSpec.scala:248]: TODO JULES

File: zio-kafka-test/src/test/scala/zio/kafka/utils/SslHelperSpec.scala [hunk 3]
```diff
@@ -236,6 +244,28 @@ object SslHelperSpec extends ZIOSpecDefaultSlf4j with KafkaRandom {
...
+              while (!interrupted) {
+                Thread.sleep(1.second.toMillis)
+                interrupted = Thread.currentThread.isInterrupted
+                println(s"interrupted: $interrupted")
+              }
+              ???
+            }(
+              settings.bootstrapServers,
+              settings.properties
+            ).fork.flatMap(fiber => ZIO.debug("INTERRUPT NOW").delay(5.seconds) *> fiber.interrupt)
```
Comment (guizmaii) [inline zio-kafka-test/src/test/scala/zio/kafka/utils/SslHelperSpec.scala:266]: TO CLEAN

**Comment by erikvanoosten** [comment]
I did a bit of research. Stackoverflow has quite a few questions about this. This is the best answer I could find: https://stackoverflow.com/questions/20752756/how-to-set-java-nio-asynchronoussocketchannel-connect-timeout Turns out you can't tell the kernel to shorten the connection timeout. 

An alternative is to use the synchronous api for which this answer is suitable: https://stackoverflow.com/a/4969788

I didn't dive into the code so treat this as a high level remark: What we can also do is race building a connection to all servers and timeout at a higher level, or does it already work like this?

**Comment by guizmaii** [comment]
@erikvanoosten Thanks for the link but I think I did find how to timeout the `SocketChannel.open` and that my current solution works. Note that my PR is still a draft so I still need to document things and clean the code.
 you shouldn't review my draft PRs ;)

**Comment by erikvanoosten** [comment]
> Note that my PR is still a draft so I still need to document things and clean the code.
> you shouldn't review my draft PRs ;)

I know. Could you create them as draft please? If the PR is marked as draft later, I automatically get a review request 😞 
afbeelding

File: zio-kafka/src/main/scala/zio/kafka/utils/SslHelper.scala [hunk 4]
```diff
@@ -102,8 +103,64 @@ object SslHelper {
...
+   *
+   * We use a `SocketChannel` to test the SSL configuration of the Kafka node. This `SocketChannel` is a resource and
+   * needs to be closed to avoid leaking memory. We want to be able to timeout the whole process if it takes too long
+   * but a `SocketChannel` is not timeoutable. So we implement a timeout mechanism that will interrupt the thread on
+   * which the `SocketChannel` is running to close it as it's one of the properties of the `SocketChannel` class.
+   *
+   * Note that there are unit-tests proving that this works.
+   *
+   * Useful links which helped to make this work:
+   *   - Discord discussion/thread with Vladimir Klyushnikov (https://github.com/vladimirkl) starting here:
```
Comment (guizmaii) [inline zio-kafka/src/main/scala/zio/kafka/utils/SslHelper.scala:157]: @vladimirkl I mentioned you here. Tell me if that ok for you or not, please 🙂

Comment (vladimirkl) [reply zio-kafka/src/main/scala/zio/kafka/utils/SslHelper.scala:157]: ok

File: zio-kafka/src/main/scala/zio/kafka/utils/SslHelper.scala [hunk 5]
```diff
@@ -115,37 +172,51 @@ object SslHelper {
...
+                case Exit.Failure(_) => ZIO.unit
+              }
+              .mapError(ConnectionError.apply)
+          )(ref.get.map {
+            case Some((channel, _)) => ZIO.attempt(channel.close()).orDie
+            case None               => ZIO.unit
+          })
+        _ <-
+          result match {
+            case None => ZIO.fail(new java.util.concurrent.TimeoutException(s"Failed to contact $address"))
```
Comment (guizmaii) [inline zio-kafka/src/main/scala/zio/kafka/utils/SslHelper.scala:208]: TODO JULES: Shouldn't this `TimeoutException` be wrapped in a `ConnectionError`? 🤔

**Pull request merged**